### ISSUE TITLE
Flaky test: TestSlowServer

### ISSUE THREAD

**Issue by jeffbean** [go-zookeeper/zk#17]
```
--- FAIL: TestSlowServer (6.05s)
    cluster_test.go:15: [ZKERR] ZooKeeper JMX enabled by default
    cluster_test.go:15: [ZKERR] Using config: /tmp/gozk591418556/srv1/zoo.cfg
    zk_test.go:908: zk: could not connect to a server
```

**Comment by pmazzini** [comment]
`TestBasicCluster` is also flaky.

```
--- FAIL: TestBasicCluster (14.82s)
##[error]    cluster_test.go:15: [ZKERR] ZooKeeper JMX enabled by default
##[error]    cluster_test.go:15: [ZKERR] ZooKeeper JMX enabled by default
##[error]    cluster_test.go:15: [ZKERR] ZooKeeper JMX enabled by default
##[error]    cluster_test.go:15: [ZKERR] Using config: /tmp/gozk258596291/srv1/zoo.cfg
##[error]    cluster_test.go:15: [ZKERR] Using config: /tmp/gozk258596291/srv3/zoo.cfg
##[error]    cluster_test.go:15: [ZKERR] Using config: /tmp/gozk258596291/srv2/zoo.cfg
##[error]    cluster_test.go:39: Create failed on node 1: zk: connection closed
```

**Comment by pmazzini** [comment]
Example TestSlowServer failure:
https://github.com/go-zookeeper/zk/runs/681501650?check_suite_focus=true

**Comment by jeffbean** [comment]
At some point we need to separate unit tests from these integration tests. The issue is we do not currently have a Fake Go Zookeeper test framework. 

Running the zk cluster is a great test, but is not ideal for unit testing.

**Comment by pmazzini** [comment]
This is what I think is happening with this flaky `TestSlowServer
` test:

* There is a [single ZooKeeper node](https://github.com/go-zookeeper/zk/blob/master/zk_test.go#L885)
* The test [closes the connection](https://github.com/go-zookeeper/zk/blob/master/zk_test.go#L897)
* On reconnect it tries to [use the same session](https://github.com/go-zookeeper/zk/blob/master/conn.go#L639)
* If the same [session expired, it returns an error](https://github.com/go-zookeeper/zk/blob/master/conn.go#L682)
* It doesn't try to establish a new session because [it already tried to connect to all existing nodes](https://github.com/go-zookeeper/zk/blob/master/conn.go#L380) (in this case only one), and it failed.
* To sum up, If the initial session is still valid, the test succeeds, if not the test fails

I think this is a legit bug in the `conn.go` logic.
